build.gradle: update to build project using Java 9 #374

Zhiyuan-Amos opened the pull request:
Fixes #374, part of #681 and #706.

Proposed merge commit message:
```
build.gradle: upgrade to use Java 9

The project currently runs on Java 8.

We should ensure compatibility with the latest public version as soon
as it is out.

Let's upgrade the project to run on Java 9.
```

damithc commented:
This fix require everyone to install Java 9?

Zhiyuan-Amos commented:
@damithc 

> In any case, we should ensure compatibility with the latest public version as soon as it is out (but we can wait a while before we make the latest version a prerequisite)

Ah sorry, I just understood what you meant here. Will make the changes accordingly.

Zhiyuan-Amos commented:
@damithc I'm not sure if we can make it compatible with both Java 8 and Java 9.

On this branch, Java 9 is supported. When running on Java 8, TestFX crashes because we have jdk9 features as dependencies which are required to support the tests when running in Java 9.

On master branch, Java 9 is unsupported because Gradle hasn't imported the jdk9 features as dependencies for tests. 

I've verified the issue from [AddressBook-Lvl2](https://github.com/nus-cs2103-AY1718S2/forum/issues/12#issuecomment-369554375) that students do not have to delete Java 9 just to run the tests successfully.

damithc commented:
I see. In that case I guess there's no choice but to do a full upgrade to Java 9.

Zhiyuan-Amos commented:
> I see. In that case I guess there's no choice but to do a full upgrade to Java 9.

Are we intending to rush this PR? An alternative solution is that when students face issues concerning Java 9 compatibility, I can respond to them.

damithc commented:
> Are we intending to rush this PR?

I guess no need.

Zhiyuan-Amos commented on build.gradle:
```diff
@@ -46,14 +46,21 @@ dependencies {
     }
-    testCompile group: 'org.testfx', name: 'openjfx-monocle', version: '1.8.0_20'
+    testRuntime 'org.testfx:testfx-internal-java9:4.0.11-alpha'
+    testRuntime 'org.testfx:openjfx-monocle:jdk-9+181'
```
Follow `testCompile` style.

damithc commented:
Shouldn't there be some updates to docs (i.e., prerequisites)?

Zhiyuan-Amos commented:
@pyokagan @yamgent @eugenepeh tagging all of you in to take a look at this small PR if you happen to be free.

Zhiyuan-Amos commented:
@eugenepeh thanks for the review! :)

Zhiyuan-Amos commented:
> What do these properties do in our project? Is there a substitute for these options? Should we be using the substitutes (if any) for our project?

Good catch. I've updated the code slightly accordingly and updated the commit message.

yamgent commented on build.gradle:
```diff
@@ -39,21 +39,29 @@ jacocoTestReport {
+    compile group: 'javax.xml.bind', name: 'jaxb-api', version: '2.2.8'
+    compile group: 'com.sun.xml.bind', name: 'jaxb-impl', version: '2.3.0'
+    compile group: 'com.sun.xml.bind', name: 'jaxb-core', version: '2.3.0'
+    compile group: 'javax.activation', name: 'activation', version: '1.1.1'
```
Who is the provider of these packages?

Also, I wonder whether it is worth to seek out alternative libraries so that we don't get a impression that we are trying to circumvent the modules system... (don't need to be in this PR)

Zhiyuan-Amos replied:
> Who is the provider of these packages?

[Java EE](https://docs.oracle.com/javaee/7/api/) for the `javax.xml.bind`, which internally uses the `com.sun.xml.bind` (that's why we have to import them as well).

> Also, I wonder whether it is worth to seek out alternative libraries so that we don't get a impression that we are trying to circumvent the modules system... (don't need to be in this PR)

Yup, we actually discussed that before in #817. :P

yamgent replied:
Ah, had a slight feeling that it was discussed before but couldn't find the exact issue, thanks for pointing it out.

Zhiyuan-Amos commented:
@Rinder5 @yamidark @vivekscl as our project has been upgraded to use Java 9 with the merging of this PR, do update your local machine's `$JAVA_HOME` to reflect this change as well, otherwise you can't compile the code locally. See this [link](http://www.baeldung.com/java-home-on-windows-7-8-10-mac-os-x-linux) and ping me up if you need any help.
